src/cs50/sql.py

```diff
@@ -41,202 +44,316 @@ def __init__(self, url, **kwargs):
 
-            def process_literal_param(self, value, dialect):
-                """Receive a literal parameter value to be rendered inline within a statement."""
+                # Remember paramstyle
+                if paramstyle is None:
+                    paramstyle = _paramstyle
 
-                def process(value):
-                    """Render a literal value, escaping as needed."""
+                # Remember placeholder's index, name
+                placeholders[index] = name
 
-                    # bool
-                    if type(value) is bool:
-                        return sqlalchemy.types.Boolean().literal_processor(dialect)(value)
+        # In case user passes args in list or tuple
+        if len(args) == 1 and (isinstance(args[0], list) or isinstance(args[0], tuple)) and len(placeholders) != 1:
+            args = args[0]
 
-                    # datetime.date
-                    elif type(value) is datetime.date:
-                        return sqlalchemy.types.String().literal_processor(dialect)(value.strftime("%Y-%m-%d"))
+        # In case user passes kwargs in dict
+        if len(args) == 1 and len(kwargs) == 0 and isinstance(args[0], dict) and len(placeholders) != 1:
+            kwargs = args[0]
 
-                    # datetime.datetime
-                    elif type(value) is datetime.datetime:
-                        return sqlalchemy.types.String().literal_processor(dialect)(value.strftime("%Y-%m-%d %H:%M:%S"))
+        # qmark
+        if paramstyle == "qmark":
 
-                    # datetime.time
-                    elif type(value) is datetime.time:
-                        return sqlalchemy.types.String().literal_processor(dialect)(value.strftime("%H:%M:%S"))
+            # Validate number of placeholders
+            if len(placeholders) < len(args):
+                raise RuntimeError("too few placeholders")
+            elif len(placeholders) > len(args):
+                raise RuntimeError("too many placeholders")
 
-                    # float
-                    elif type(value) is float:
-                        return sqlalchemy.types.Float().literal_processor(dialect)(value)
+            # Escape values
+            for i, index in enumerate(placeholders.keys()):
+                tokens[index] = self._escape(args[i])
 
-                    # int
-                    elif type(value) is int:
-                        return sqlalchemy.types.Integer().literal_processor(dialect)(value)
+        # numeric
+        elif paramstyle == "numeric":
 
-                    # long
-                    elif sys.version_info.major != 3 and type(value) is long:
-                        return sqlalchemy.types.Integer().literal_processor(dialect)(value)
+            # Escape values
+            for index, name in placeholders.items():
+                i = int(name) - 1
+                if i < 0 or i >= len(args):
+                    raise RuntimeError("placeholder out of range")
```

Comment:
Should we mention the out of range index here? And do we not care about extra args?

Reply:
Done.

As for extra args, I ban them for positional placeholders, since, e.g.,

```
SELECT * FROM foo WHERE bar = ?", 1, 2)
```

is presumably user error. 

However, it seems a bit extreme to prevent the user from passing in a `dict` with more keys than are needed. We could prevent the user from passing in more `kwargs` than are needed, but I'm currently treating `dict` and `kwargs` as equivalent, so I wonder if it's best to let the user do, e.g.

```
SELECT * FROM foo WHERE bar = :bar", bar=1, baz=2
```

as well? Or should we reject altogether passing more keys than are needed, whether in a `dict` or in `kwargs`?

Reply:
If we take it for a reference, it seems SQLAlchemy doesn't raise an exception when there are more kwargs than needed so I guess we could leave it as is.

fwiw though, [one of my test cases](https://github.com/cs50/python-cs50/blob/3.2/tests/sql.py#L258) that I (commented out temporarily to ask you) doesn't raise a `RuntimeError` with extra numeric args, while works as expected with qmark. All the other commented out test cases fail as well, but I take it that's expected per your other comment on supporting `IN`.

Reply:
PS, SQLAlchemy does raise an exception when there are more numeric args than needed it seems.

Reply:
Yeah, I'm currently treating `:N` like a named `:name` argument, though we could just clamp down on those too. Lemme ponder.

Reply:
Clamped.